[CTI] Enable IM Rule Preview

ecezalp opened the pull request:
## Summary

Enables Indicator Match Rule Preview, with a 60 second timeout. 

60 seconds is used as a placeholder and is pending product confirmation - but any updates to the duration can be made in a subsequent PR.

## Testing

- test the preview functionality as expected (by posting matching events and indicators to two separate indices using devtools, and confirm that the preview matrix histogram populates as expected on the preview UI)
- to test the timeout functionality,  update `x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/preview_rules_route.ts` line 51 to `const PREVIEW_TIMEOUT_SECONDS = 0.1;`
- integration test coverage will be added with https://github.com/elastic/security-team/issues/708 (chatted with Davis)

### Checklist

Delete any items that are not applicable to this PR.

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

nkhristinin reviewed (approved):
Tested locally, it works as expected!

<img width="934" alt="Screenshot 2022-03-02 at 15 33 59" src="https://user-images.githubusercontent.com/7609147/156384473-5b849ec1-6b59-49f6-b2ce-a74023caeff9.png">

marshallmain commented on x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/preview_rules_route.ts:
```diff
@@ -224,6 +226,14 @@ export const previewRulesRoute = async (
+            logs.push({
+              warnings: [`Preview timed out after ${PREVIEW_TIMEOUT_SECONDS} seconds`],
+              errors: [],
+              startedAt: moment().toDate().toISOString(),
```
The `logs` structure contains an array of objects where each object corresponds to a particular rule invocation, so adding an object that is not from a rule invocation mixes different concepts together. Can we return this warning as a separate object in the response body instead of including it in `logs`?

marshallmain reviewed (commented):
It would be useful, now that the preview can time out, to include the duration of each rule invocation in the log object. This would give better insight into why the preview times out (e.g. one invocation is very long vs all invocations are somewhat long)

ecezalp commented:
latest changes

- fixed a bug where errors were not showing up on the UI
  - this was caused by the logged items on the statusChanges array on the previewLogExecutionClient not properly transferring to the route function context
- added `isAborted` to previewRouteResponse as discussed in Marshall's [comment](https://github.com/elastic/kibana/pull/126651#discussion_r817922967)
- added rule execution duration to frontend [comment](https://github.com/elastic/kibana/pull/126651#pullrequestreview-897904425)
- stopped subsequent preview rule executions when an error is found.

screenshots
preview no hits, no errors, no warnings
<img width="600" alt="Screen Shot 2022-03-03 at 3 41 54 PM" src="https://user-images.githubusercontent.com/18617331/156657016-94edd8ff-bfb9-438a-b32e-3c39c29656e5.png">

preview hits, no errors, no warnings
<img width="600" alt="Screen Shot 2022-03-03 at 4 00 19 PM" src="https://user-images.githubusercontent.com/18617331/156657017-5f442a43-d924-4f4b-9a7d-f68cac6c8e80.png">

preview timeout warning
<img width="600" alt="Screen Shot 2022-03-03 at 3 45 37 PM" src="https://user-images.githubusercontent.com/18617331/156657018-9fc1307d-4e36-4cf4-bd9d-0dd5d4072c0e.png">

preview error with duration
<img width="600" alt="Screen Shot 2022-03-03 at 3 42 09 PM" src="https://user-images.githubusercontent.com/18617331/156657019-bc7c211d-279d-45a2-9f6c-532c6a3c767c.png">

ecezalp commented:
@elasticmachine merge upstream

ecezalp commented:
@elasticmachine merge upstream

marshallmain reviewed (approved):
LGTM 🚀

kibanamachine commented:
## ⚪ Backport skipped
The pull request was not backported as there were no branches to backport to. If this is a mistake, please apply the desired version labels or run the backport tool manually.

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 126651
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)
<!--- Backport version: 7.4.0 -->

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create backports run `node scripts/backport --pr 126651` or prevent reminders by adding the `backport:skip` label.

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create backports run `node scripts/backport --pr 126651` or prevent reminders by adding the `backport:skip` label.
